FirebaseSessions/Sources/SettingsDownloadClient.swift

```diff
@@ -0,0 +1,99 @@
+//
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+@_implementationOnly import GoogleUtilities
+
+protocol SettingsDownloadClient {
+  func fetch(completion: @escaping (Result<[String: Any], Error>) -> Void)
+}
+
+class SettingsDownloader: SettingsDownloadClient {
+  private let appInfo: ApplicationInfoProtocol
+  private let identifiers: IdentifierProvider
+  private let installations: InstallationsProtocol
+
+  init(appInfo: ApplicationInfoProtocol, identifiers: IdentifierProvider,
+       installations: InstallationsProtocol) {
+    self.appInfo = appInfo
+    self.identifiers = identifiers
+    self.installations = installations
+  }
+
+  func fetch(completion: @escaping (Result<[String: Any], Error>) -> Void) {
+    guard let url = url else {
+      completion(.failure(FirebaseSessionsError.SettingsError("Invalid URL")))
+      return
+    }
+
+    buildRequest(url: url) { result in
+      switch result {
+      case let .success(request):
+        let task = URLSession.shared.dataTask(with: request) { data, response, error in
+          if let data = data {
+            if let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
+              completion(.success(dict))
+            } else {
+              completion(.failure(FirebaseSessionsError
+                  .SettingsError("Failed to parse JSON to dictionary")))
+            }
+          } else if let error = error {
+            completion(.failure(FirebaseSessionsError.SettingsError(error.localizedDescription)))
+          }
+        }
+        task.resume()
+      case let .failure(error):
+        completion(.failure(FirebaseSessionsError.SettingsError(error.localizedDescription)))
+      }
+    }
+  }
+
+  private var url: URL? {
+    var components = URLComponents()
+    components.scheme = "https"
+    components.host = "firebase-settings.crashlytics.com"
+    components.path = "/spi/v2/platforms/android/gmp/\(appInfo.appID)/settings"
+    components.queryItems = [
+      URLQueryItem(name: "build_version", value: appInfo.appBuildVersion),
+      URLQueryItem(name: "display_version", value: appInfo.appDisplayVersion),
+    ]
+    return components.url
+  }
+
+  private func buildRequest(url: URL, completion: @escaping (Result<URLRequest, Error>) -> Void) {
+    installations.installationID { [self] result in
```

Comment:
I might recommend pulling this out into the `fetch` method. This may simplify our code so we can make `buildRequest` synchronous and have it just return a value.  It might also be a bit more readable - I had not envisioned `buildRequest` would be getting FIIDs so moving FIID to `fetch` kindof puts it in our face more so to speak so we stay aware of it.

Reply:
I'm not entirely sure what you mean, is it something like 
```
func fetch(...) {
  guard let url = url else {
    return
  }
  
  installations.installationID { result in
    switch result {
    case let .success(fiid):
      let request = buildRequest(url: url, installationID: fiid)
      let task = URLSession.shared.dataTask(with: request) { data, response, error in
          if let data = data {
            if let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
              completion(.success(dict))
            } else {
              completion(.failure(FirebaseSessionsError
                  .SettingsError("Failed to parse JSON to dictionary")))
            }
          } else if let error = error {
            completion(.failure(FirebaseSessionsError.SettingsError(error.localizedDescription)))
          }
      }
      task.resume()
    case let .failure(error): 
      completion(.failure(FirebaseSessionsError.SettingsError(error.localizedDescription)))
    }
  }
}
```

Reply:
Yes this is exactly it 👍 To me the advantage is we've moved all our async code to one spot so it's easier to see the different paths to calling `completion`

Reply:
Done 👍